src/Tasks/Copy.cs

```diff
@@ -299,7 +299,7 @@ private void TryCopyViaLink(string linkComment, MessageImportance messageImporta
             // physical file on disk.
             // Since we'll fall back to a copy (below) this will fail and issue a correct
             // message in the case that the source and destination are in fact the same file.
-            if (destinationFileExists && !IsMatchingSizeAndTimeStamp(sourceFileState, destinationFileState))
+            if (destinationFileExists && !PathsAreIdentical(sourceFileState, destinationFileState))
```

Comment:
I don't think this fixes the problem mentioned in #3972. In #3972 the error seems to come from the fact that the destination file is open in another process, not from the fact that this line decided not to delete (which is only a message, not a warning or error). So before attempting a fix, you should identify what other process has the file opened.

Can you please elaborate why you think this fixes the issue?

Also, your fix will almost always result in deleting the destination, which is not the intent of the comment above (cannot create links over existing files, so delete the destination if the files are not the same, otherwise fallback to good old copy). This is because in a copy the source and destination paths are probably always different, so at this point you might as well keep only the `destinationFileExists` check.

Reply:
@cdmihai Thanks for looking at this PR.
I'm pretty sure that my change fixes #3972, explanation is following.
If you publish "netcoreapp", it does copy operation with `CopyAlways` / `SkipUnchangedFiles=false`. If you combine `SkipUnchangedFiles=false` with hard or symbolic  linking and you try to run that operation twice, you get an error reported in #3972. Although the error message says "The process cannot access the file ... because it is being used by another process", the actual reason for that error is because `msbuild` tries to copy file over existing hardlink (or symbolic link) and there is no other proces locking the file.
To reproduce #3972, you have to run publish twice. First publish succeeds and create hard (or symbolic) link. It fails second time because:
- 'msbuild' doesn't delete link as `IsMatchingSizeAndTimeStamp(sourceFileState, destinationFileState)` is `true`
- link operation fails, because destination file already exists
- 'msbuild' tries to copy file over existing link. That copy operation fails with error message "The process cannot access the file ... because it is being used by another process" and HResult=0x80070020`.

My change, fixes it, because we will always delete destination before creating a link. Therefore it succeeds and then we don't trigger an issue with copying file over existing link.

I thought that original code was trying to prevent from deleting the only copy of given file, if someone would try to copy a file over itself. After looking a little bit more, in case,where source and destination are the same, `TryCopyViaLink` is not even called. Therefore we can keep only the `destinationFileExists` check.

Reply:
Ah, I see, so there's maybe three problems:
1. msbuild incorrectly reports that the file is opened in another process when in fact the symlink creation failed
2. [IsMatchingSizeAndTimeStamp](https://github.com/Microsoft/msbuild/blob/d8d43bca5bfe28e62c22211ddd1c5578cf3e37f8/src/Tasks/Copy.cs#L141) might be reading the link's file attributes instead of the link's target's file attributes. The [documentation](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-getfileattributesexa) says `Symbolic link behavior—If the path points to a symbolic link, the function returns attributes for the symbolic link.` If this is so, then the SkipUnchanged check will always fail when comparing real file to linked file.
3. TryCopyViaLink is failing with an exception (**please confirm**, is it an exception or does it just return false?) when the destination already exists. In my view, the behaviour should be to do the size and timestamp check (by following the link) and if unchanged, do nothing and return true so the caller does not real copy it. If changed, then delete and re-link. My reasoning here is that the symlink option is intended for performance, so doing the size/timestamp check even when `SkipUnchangedFiles=false`  is alright. @rainersigwald for opinion :)

Reply:
Ad 2. Ah, that is important observation, I confirmed it with debugger, it is exactly as you say. When we compare real file and linked file:
- for hard links `IsMatchingSizeAndTimeStamp` returns true
- for symbolic links  `IsMatchingSizeAndTimeStamp` returns false

Ad 1.  Just for clarification, #3972 occurs only for hardlinks. In case of symbolic links, it is mitigated by  the fact that `IsMatchingSizeAndTimeStamp` reads the link's file attributes instead of the link's target's file attribute.

3. `TryCopyViaLink` doesn't throw an exception, it just return false if link cannot be created. Actual exception, is thrown by [File.Copy](https://github.com/Microsoft/msbuild/blob/d13c0ebd8f7a1fe88fcc86253011a9688f958dc2/src/Tasks/Copy.cs#L269), when destination exists and is a link.
I'm not convinced that we should sacrifice correctness in favour of performance. I would rather choose to always delete and recreate a link when `SkipUnchangedFiles=false`.

Reply:
Mkay, so there's two potential fixes:
1. Make `IsMatchingSizeAndTimeStamp` traverse the links to get the real file's attributes. This also fixes `SkipUnchnaged=true` for symlinks. But always checking for symlinks might have perf implications, so maybe special-casing the symlink behaviour only in symlink contexts.
2. Just delete and re-symlink, no checks.

I'll let you choose the solution. I have a weak preference for the first one. You could use [benchmarkdotnet](https://benchmarkdotnet.org/) to see what's more expensive, following symlinks in a loop to get the real attributes, or deleting and recreating.

Reply:
After talking to @rainersigwald , fix 1. also raises the problem of what the right behaviour should be, in terms of the nature of the source and destination.

If source is a real file, and destination is a symlink, then I think the link should be followed to enable skipping the copy.

If both are symlinks, then it gets ambiguous. The two links might even point to different files, in which case just recreating seems fine. If they point to the same file, then maybe the users want to actually recreate the link when `SkipUnchanged=false` to get the symlink's timestamp updated.

To avoid this headache, option 2 seems more favorable, as its simpler, and it's unknown what the perfect impact of recreating the link is, as opposed to doing the follow-through check. Should be minimal on ssd drives.

Reply:
FWIW, I've also found out that traversing symlinks comes with a performance cost that is not negligible (https://github.com/Microsoft/msbuild/blob/b630e674c729787d57a485ef5338915643eb1cea/src/Shared/NativeMethodsShared.cs#L1002). 

Option 2 ("just delete and re-symlink, no checks") is fine for me.
If there are no other concerns, then I suppose we can merge this PR ?